velox/functions/sparksql/specialforms/FromJson.cpp

```diff
@@ -0,0 +1,571 @@
+      auto& keyType = writer.type()->childAt(0);
+      auto& valueType = writer.type()->childAt(1);
+      SIMDJSON_ASSIGN_OR_RAISE(auto object, value.get_object());
+      for (auto fieldResult : object) {
+        SIMDJSON_ASSIGN_OR_RAISE(auto field, fieldResult);
+        SIMDJSON_ASSIGN_OR_RAISE(auto key, field.unescaped_key(true));
+        // If casting to map of JSON values, nulls in map values should become
+        // the JSON text "null".
+        if (field.value().is_null()) {
+          writerTyped.add_null().castTo<Varchar>().append(key);
+        } else {
+          auto writers = writerTyped.add_item();
+          std::get<0>(writers).castTo<Varchar>().append(key);
+          SIMDJSON_TRY(VELOX_DYNAMIC_TYPE_DISPATCH(
+              ParseJsonTypedImpl<simdjson::ondemand::value>::apply,
+              valueType->kind(),
+              field.value(),
+              std::get<1>(writers),
+              false));
+        }
+      }
+      return simdjson::SUCCESS;
+    }
+  };
+
+  template <typename Dummy>
+  struct KindDispatcher<TypeKind::ROW, Dummy> {
+    static simdjson::error_code
+    apply(Input value, exec::GenericWriter& writer, bool isRoot) {
+      auto& rowType = writer.type()->asRow();
+      auto& writerTyped = writer.castTo<DynamicRow>();
+      if (value.type().error() != ::simdjson::SUCCESS) {
+        writerTyped.set_null_at(0);
+        return simdjson::SUCCESS;
+      }
+      auto type = value.type().value_unsafe();
+      if (type == simdjson::ondemand::json_type::object) {
+        SIMDJSON_ASSIGN_OR_RAISE(auto object, value.get_object());
+
+        folly::F14FastMap<std::string, int32_t> fieldIndices;
+        const auto size = rowType.size();
+        for (auto i = 0; i < size; ++i) {
+          std::string key = rowType.nameOf(i);
+          fieldIndices[key] = i;
+        }
+
+        std::string key;
+        for (auto fieldResult : object) {
+          if (fieldResult.error() != ::simdjson::SUCCESS) {
+            continue;
+          }
+          auto field = fieldResult.value_unsafe();
+          if (!field.value().is_null()) {
+            SIMDJSON_ASSIGN_OR_RAISE(key, field.unescaped_key(true));
+
+            auto it = fieldIndices.find(key);
+            if (it != fieldIndices.end()) {
+              const auto index = it->second;
+              it->second = -1;
+
+              auto res = VELOX_DYNAMIC_TYPE_DISPATCH(
+                  ParseJsonTypedImpl<simdjson::ondemand::value>::apply,
+                  rowType.childAt(index)->kind(),
+                  field.value(),
+                  writerTyped.get_writer_at(index),
+                  false);
+              if (res != simdjson::SUCCESS) {
+                writerTyped.set_null_at(index);
+              }
+            }
+          }
+        }
+
+        for (const auto& [key, index] : fieldIndices) {
+          if (index >= 0) {
+            writerTyped.set_null_at(index);
+          }
+        }
+      } else {
```

Comment:
What are the 'else' cases, perhaps add some comments.

Reply:
It's for handling other json types.Although they are invalid, behavior differs based on isRoot.